ray-operator/controllers/ray/expectations/active_expectation.go

```diff
@@ -0,0 +1,382 @@
+	Delete(key string) error
+}
+
+func NewActiveExpectations(client client.Client) ActiveExpectationsInterface {
+	return &ActiveExpectations{
+		Client:   client,
+		subjects: cache.NewStore(ExpKeyFunc),
+	}
+}
+
+type ActiveExpectations struct {
+	client.Client
+	subjects cache.Store
+}
+
+func (ae *ActiveExpectations) ExpectCreate(key string, kind ExpectedResourceType, namespace, name string) error {
+	return ae.expectCreateOrDelete(key, kind, Create, namespace, name)
+}
+
+func (ae *ActiveExpectations) ExpectDelete(key string, kind ExpectedResourceType, namespace, name string) error {
+	return ae.expectCreateOrDelete(key, kind, Delete, namespace, name)
+}
+
+func (ae *ActiveExpectations) expectCreateOrDelete(key string, kind ExpectedResourceType, action ActiveExpectationAction, namespace, name string) error {
+	expectation, exist, err := ae.subjects.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get expectation for active expectations %s when expecting: %s", key, err)
+	}
+
+	if !exist {
+		expectation = NewActiveExpectation(ae.Client, key, namespace)
+		if err = ae.subjects.Add(expectation); err != nil {
+			return err
+		}
+	}
+
+	if err = expectation.(*ActiveExpectation).expectCreateOrDelete(kind, name, action); err != nil {
+		return fmt.Errorf("fail to expect %s/%s for action %d: %s", kind, name, action, err)
+	}
+
+	return nil
+}
+
+func (ae *ActiveExpectations) ExpectUpdate(key string, kind ExpectedResourceType, namespace, name string, updatedResourceVersion string) error {
+	rv, err := strconv.ParseInt(updatedResourceVersion, 10, 64)
+	if err != nil {
+		panic(fmt.Sprintf("fail to parse resource version %s of resource %s/%s to int64 for subject %s: %s",
+			updatedResourceVersion, kind, name, key, err))
+	}
+
+	if _, exist := ResourceInitializers[kind]; !exist {
+		panic(fmt.Sprintf("kind %s is not supported for Active Expectation", kind))
+	}
+
+	expectation, exist, err := ae.subjects.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get expectation for active expectations %s when expecting: %s", key, err)
+	}
+
+	if !exist {
+		expectation = NewActiveExpectation(ae.Client, namespace, key)
+		if err := ae.subjects.Add(expectation); err != nil {
+			return err
+		}
+	}
+
+	if err := expectation.(*ActiveExpectation).expectUpdate(kind, name, rv); err != nil {
+		return fmt.Errorf("fail to expect %s/%s for action %d, %s: %s", kind, name, Update, updatedResourceVersion, err)
+	}
+
+	return nil
+}
+
+func (ae *ActiveExpectations) IsSatisfied(key string) (satisfied bool, err error) {
+	expectation, exist, err := ae.subjects.GetByKey(key)
+	if err != nil {
+		return false, fmt.Errorf("fail to get expectation for active expectations %s when check satisfication: %s", key, err)
+	}
+
+	if !exist {
+		return true, nil
+	}
+
+	defer func() {
+		if satisfied {
+			ae.subjects.Delete(expectation)
+		}
+	}()
+
+	satisfied, err = expectation.(*ActiveExpectation).isSatisfied()
```

Comment:
There are many read-after-write operations in the `ActiveExpectations`. Should we use a mutex to wrap these operations? For example, will the above `ae.subjects.Delete(expectation)` delete an unsatisfied expectation?

Reply:
The subjects of `ActiveExpectations` are `ThreadSafeStore`, utilizing `store` provided by [k8s.io/client-go/tools/cache](https://github.com/kubernetes/client-go/blob/master/tools/cache/store.go#L282). Therefore, operations on `ActiveExpectations.subjects` are thread-safe.
For `ActiveExpectations.subjects` items, the `ActiveExpectation` also utilizes a `ThreadSafeStore`.

Reply:
👍